### ISSUE TITLE
Most efficient way to get Numpy array from reader?

### ISSUE THREAD

**Issue by jfainberg** [pykaldi/pykaldi#45]
Hi,

I want to read in integer vectors and manipulate them as Numpy arrays. I noticed that if I use rxfilenames and use kaldi.util.io.read_vector I get a Vector object on which I can call numpy() efficiently. But if I use SequentialIntVectorReader, the value returned is a list. Would calling Vector on that list copy the underlying memory? What's the most efficient way to handle this use case?

Thanks for writing PyKaldi!

**Comment by dogancan** [comment]
Hi Joachim

TLDR -- Unfortunately, there is not a more efficient way of doing this particular conversion. In your case, the best way is to directly convert the list to a NumPy array.

#### Why does `SequentialIntVectorReader` returns a `list` instead of a `Vector`?

Unlike NumPy `ndarray` type, Kaldi `Vector/Matrix` types are limited to single/double precision floating point data types. Kaldi uses `std::vector` type to store vectors of other data types. When you read an integer vector from file, Kaldi constructs a `std::vector<int>` to hold the data and CLIF, the wrapper generator we are using, converts C++ `std::vector` objects to Python `list` objects.

#### Would calling `Vector` on a `list` copy the underlying memory?

Since Python lists are heterogeneous containers that do not store their elements in a continuous memory region, they cannot be converted to PyKaldi vectors or Numpy arrays by sharing the underlying memory. In fact, all these types of conversions from arbitrary iterables to PyKaldi types are handled by constructing NumPy arrays under the hoods. When you call `Vector` on a Python `list`, first it will be converted to a temporary NumPy `ndarray` by allocating new memory and **copying** the items in the list and then the array will be converted into a PyKaldi `Vector` by allocating yet another chunk of memory and **copying** the items in the array. So, this operation will copy the contents of the input list **twice**. The reason why the second copy is needed is rather complicated. The "proper" way of sharing the data owned by a NumPy array is by constructing a `SubVector/SubMatrix` object.

**Comment by jfainberg** [comment]
Thanks so much for the clear and informative answer, @dogancan !
